hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -254,64 +293,105 @@ private void handlePlatformTransaction(
+                recordBuilder.status(preCheckResult.responseCodeEnum());
 
-            // TODO: Finalize transaction with the help of the token service
-            final var finalizationContext =
-                    new FinalizeContextImpl(preHandleResult.payer(), recordBuilder, configuration, stack);
-            transactionFinalizer.finalizeParentRecord(
-                    finalizationContext, List.of()); // TODO Need actual list of child records?
+            } else {
+                feeAccumulator.charge(payer, fees);
+                try {
+                    // Dispatch the transaction to the handler
+                    dispatcher.dispatchHandle(context);
+
+                    // commit state
+                    stack.commitFullStack();
+                    recordBuilder.status(SUCCESS);
+                } catch (final HandleException e) {
+                    rollback(e.getStatus(), stack, recordListBuilder);
+                    feeAccumulator.charge(payer, fees);
+                }
+            }
+        } catch (final Exception e) {
+            logger.error("An unexpected exception was thrown during handle", e);
+            rollback(ResponseCodeEnum.FAIL_INVALID, stack, recordListBuilder);
+            if (payer != null && fees != null) {
+                feeAccumulator.charge(payer, fees);
+            }
+        }
 
-            recordBuilder.status(SUCCESS);
+        transactionFinalizer.finalizeParentRecord(payer, tokenServiceContext);
 
-            // commit state
-            stack.commitFullStack();
+        // store all records at once
+        final var recordListResult = recordListBuilder.build();
+        recordCache.add(creator.nodeId(), payer, recordListResult.mainRecord().record(), consensusNow);
+        blockRecordManager.endUserTransaction(recordListResult.recordStream(), state);
 
-            // Notify responsible facility if system-file was uploaded
-            systemFileUpdateFacility.handleTxBody(state, txBody);
+        return txBody;
+    }
 
-        } catch (final PreCheckException e) {
-            recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
-        } catch (final HandleException e) {
-            recordFailedTransaction(e.getStatus(), recordBuilder, recordListBuilder);
-        } catch (final Throwable e) {
-            logger.error("An unexpected exception was thrown during handle", e);
-            recordBuilder.status(ResponseCodeEnum.UNKNOWN);
-        }
+    @NonNull
+    private FeeAccumulator createFeeAccumulator(
+            @NonNull final SavepointStackImpl stack,
+            @NonNull final Configuration configuration,
+            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder) {
+        final var serviceApiFactory = new ServiceApiFactory(stack, configuration);
+        final var tokenApi = serviceApiFactory.getApi(TokenServiceApi.class);
+        return new FeeAccumulatorImpl(tokenApi, recordBuilder);
+    }
```

Comment:
I'm not sure if it is worth writing an entire method for a piece of code that is used in only one place. However, it is probably inlined, so it probably has no downside.